### PR TITLE
More types for synapse.util, part 1

### PR DESCRIPTION
**Pull request by DMRobertson**
I noticed the `no-untyped-defs` option in mypy.ini and saw that we weren't using this in the recently annotated `synapse.util`. I thought I'd have a go. This is a first batch of changes to make various files pass `no-untyped-defs`. In a few places I found myself blocked because I didn't understand how `synapse.metrics` was annotated, but @squahtx's work in #10847 will help a lot here.

Should be reviewable commit-by-commit.

### CODE REVIEW

**Comment by DMRobertson** [comment]
Looks like the LRU cache changes conflict with my recent changes in #10829.

File: synapse/util/caches/response_cache.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ def _set(
         key = context.cache_key
         self.pending_result_cache[key] = result
 
-        def on_complete(r):
+        def on_complete(r: Any) -> Any:
```
Comment (squahtx) [inline synapse/util/caches/response_cache.py:129]: nit: `def on_complete(r: T) -> T` would be more precise

File: synapse/util/patch_inline_callbacks.py [hunk 2]
```diff
@@ -84,7 +86,7 @@ def wrapped(*args, **kwargs):
                 print(err, file=sys.stderr)
                 raise Exception(err)
 
-            def check_ctx(r):
+            def check_ctx(r: Any) -> Any:
```
Comment (squahtx) [inline synapse/util/patch_inline_callbacks.py:89]: `def check_ctx(r: T) -> T` would be more precise

File: synapse/util/patch_inline_callbacks.py [hunk 3]
```diff
@@ -37,12 +37,14 @@ def do_patch() -> None:
     if _already_patched:
         return
 
-    def new_inline_callbacks(f):
+    def new_inline_callbacks(f: Callable[..., Generator]) -> Callable[..., Any]:
```
Comment (squahtx) [inline synapse/util/patch_inline_callbacks.py:40]: In theory this ought to have the same type signature as the original `defer.inlineCallbacks`.
Perhaps we can borrow that?

https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/defer.py#L1827-L1831 :
```py
def inlineCallbacks(f: Callable[..., Generator[Deferred[object], object, T]]) -> Callable[..., Deferred[T]]:
```

which would imply that:
  * `wrapped` below returns a `Deferred[T]`
  * `orig` is a `Callable[..., Deferred[T]]`
  * `_check_yield_points` accepts a `Callable[..., Generator[Deferred[object], object, T]]`
  * and I think `check_yield_points_inner` might return a `Generator[Deferred[object], object, T]`

File: synapse/util/caches/response_cache.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/util/caches/response_cache.py:129]: Agreed. I think we might be able to go further though: I think `r: RV`. Reasoning:

- `wrap`'s callback returns an `Awaitable[RV]`

https://github.com/matrix-org/synapse/blob/894be5fb9a9b6fa4d5ef3d4daf1926975d26b303/synapse/util/caches/response_cache.py#L148-L155

- That callback is `run_in_background` to build a `Deferred`, which is passed to `_set`

https://github.com/matrix-org/synapse/blob/894be5fb9a9b6fa4d5ef3d4daf1926975d26b303/synapse/util/caches/response_cache.py#L200-L201

- That deferred's result is what gets passed to on_complete

File: synapse/util/patch_inline_callbacks.py (hunk 3 above)
Comment (DMRobertson) [reply synapse/util/patch_inline_callbacks.py:40]: Thanks for this. (I struggled to grok this and clearly just phoned it in with `Any`).

Logic seems sound and mypy seems happy with those annotations!

File: synapse/util/metrics.py [hunk 4]
```diff
@@ -61,10 +63,15 @@
     sub_metrics=["real_time_max", "real_time_sum"],
 )
 
-T = TypeVar("T", bound=Callable[..., Any])
+R = TypeVar("R")
+F = Callable[..., R]
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/util/metrics.py:67]: I think so. I wonder why mypy didn't spot this? (I guess I force it with a `cast` below, but even removing the cast it was happy.)

File: synapse/util/caches/deferred_cache.py [hunk 5]
```diff
@@ -282,11 +283,11 @@ def eb(_fail) -> None:
 
     def prefill(
         self, key: KT, value: VT, callback: Optional[Callable[[], None]] = None
-    ):
+    ) -> None:
         callbacks = [callback] if callback else []
         self.cache.set(key, value, callbacks=callbacks)
 
-    def invalidate(self, key):
+    def invalidate(self, key: KT) -> None:
```
Comment (erikjohnston) [inline synapse/util/caches/deferred_cache.py:290]: I'm not sure this actually accurate when using `TreeCache` and tuple keys, as I think we support giving a subset of the keys to delete multiple entries at once. I don't know how we type that though, so I'm tempted to just leave it for now.

File: synapse/util/caches/lrucache.py [hunk 6]
```diff
@@ -86,7 +87,8 @@ def _get_size_of(val: Any, *, recurse=True) -> int:
 T = TypeVar("T")
 
 
-def enumerate_leaves(node, depth):
+# TODO I think this is this unused?
```
Comment (erikjohnston) [inline synapse/util/caches/lrucache.py:90]: Looks like it yeah, let's delete it.

File: synapse/util/metrics.py [hunk 7]
```diff
@@ -82,16 +89,16 @@ async def foo(...):
 
     """
 
-    def wrapper(func: T) -> T:
+    def wrapper(func: F) -> F:
```
Comment (erikjohnston) [inline synapse/util/metrics.py:92]: Why this change? We lose a bunch of typing information here, and isn't what the mypy docs recommend for annotating decorators.

File: synapse/util/versionstring.py [hunk 8]
```diff
@@ -31,11 +35,11 @@ def get_version_string(module) -> str:
...
     """
 
-    cached_version = getattr(module, "_synapse_version_string_cache", None)
-    if cached_version:
+    cached_version = version_cache.get(module)
+    if cached_version is not None:
         return cached_version
 
-    version_string = module.__version__
+    version_string = getattr(module, "__version__", "<unknown>")
```
Comment (erikjohnston) [inline synapse/util/versionstring.py:42]: I'd vote we keep the behaviour where we raise an exception if the version is unknown. Mainly because let's not change behaviour in a PR that adds type hints.

File: synapse/util/metrics.py (hunk 7 above)
Comment (DMRobertson) [reply synapse/util/metrics.py:92]: As in [this](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#decorators)? Hadn't seen this before.

And I don't think I'd properly grokked the difference between a bound typevar and just using the bound verbatim.

----
Motivation: I wanted to encode in the type system that the `func` being decorated was asynchronous, and so too was the decorated version `measured_func`. Maybe there's a better way? Or maybe it's not worth it?

File: synapse/util/versionstring.py (hunk 8 above)
Comment (DMRobertson) [reply synapse/util/versionstring.py:42]: That makse sense. I got tunnel vision trying to make mypy pass.

File: synapse/util/caches/deferred_cache.py (hunk 5 above)
Comment (DMRobertson) [reply synapse/util/caches/deferred_cache.py:290]: Agreed, let's ditch it

File: synapse/util/metrics.py (hunk 7 above)
Comment (DMRobertson) [reply synapse/util/metrics.py:92]: I wanted to emphasise that the function being decorated should be asynchronous. But I'm happy to revert this, I don't think is a particularly good job

File: synapse/util/versionstring.py (hunk 8 above)
Comment (DMRobertson) [reply synapse/util/versionstring.py:42]: Fair enough. Are you happy to keep the cache in a dictionary in this module (as opposed to adding an attribute to each module?)

**Review by squahtx — APPROVED** [review]
LGTM once CI is happy

File: synapse/util/caches/ttlcache.py [hunk 9]
```diff
@@ -159,12 +159,12 @@ def expire(self) -> None:
             del self._expiry_list[0]
 
 
-@attr.s(frozen=True, slots=True)
-class _CacheEntry:
```
Comment (DMRobertson) [inline synapse/util/caches/ttlcache.py:163]: CI is unhappy here.

<details>

```
2021-09-27T15:54:47.4244222Z [ERROR]
2021-09-27T15:54:47.4244683Z Traceback (most recent call last):
2021-09-27T15:54:47.4251489Z   File "/home/runner/work/synapse/synapse/.tox/py/lib/python3.6/site-packages/twisted/trial/runner.py", line 693, in loadByName
2021-09-27T15:54:47.4252855Z     return self.suiteFactory([self.findByName(name, recurse=recurse)])
2021-09-27T15:54:47.4254696Z   File "/home/runner/work/synapse/synapse/.tox/py/lib/python3.6/site-packages/twisted/trial/runner.py", line 453, in findByName
2021-09-27T15:54:47.4255759Z     obj = reflect.namedModule(searchName)
2021-09-27T15:54:47.4257137Z   File "/home/runner/work/synapse/synapse/.tox/py/lib/python3.6/site-packages/twisted/python/reflect.py", line 157, in namedModule
2021-09-27T15:54:47.4258112Z     topLevel = __import__(name)
2021-09-27T15:54:47.4258979Z   File "/home/runner/work/synapse/synapse/tests/rest/client/test_third_party_rules.py", line 27, in <module>
2021-09-27T15:54:47.4259789Z     from tests import unittest
2021-09-27T15:54:47.4260567Z   File "/home/runner/work/synapse/synapse/tests/unittest.py", line 38, in <module>
2021-09-27T15:54:47.4261634Z     from synapse.federation.transport import server as federation_server
2021-09-27T15:54:47.4262824Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/server/__init__.py", line 21, in <module>
2021-09-27T15:54:47.4264029Z     from synapse.federation.transport.server._base import (
2021-09-27T15:54:47.4265283Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/server/_base.py", line 31, in <module>
2021-09-27T15:54:47.4266216Z     from synapse.server import HomeServer
2021-09-27T15:54:47.4267084Z   File "/home/runner/work/synapse/synapse/synapse/server.py", line 54, in <module>
2021-09-27T15:54:47.4268186Z     from synapse.federation.federation_client import FederationClient
2021-09-27T15:54:47.4275541Z   File "/home/runner/work/synapse/synapse/synapse/federation/federation_client.py", line 58, in <module>
2021-09-27T15:54:47.4277013Z     from synapse.federation.transport.client import SendJoinResponse
2021-09-27T15:54:47.4278483Z   File "/home/runner/work/synapse/synapse/synapse/federation/transport/client.py", line 33, in <module>
2021-09-27T15:54:47.4280832Z     from synapse.http.matrixfederationclient import ByteParser
2021-09-27T15:54:47.4282277Z   File "/home/runner/work/synapse/synapse/synapse/http/matrixfederationclient.py", line 67, in <module>
2021-09-27T15:54:47.4283684Z     from synapse.http.federation.matrix_federation_agent import MatrixFederationAgent
2021-09-27T15:54:47.4285068Z   File "/home/runner/work/synapse/synapse/synapse/http/federation/matrix_federation_agent.py", line 41, in <module>
2021-09-27T15:54:47.4286572Z     from synapse.http.federation.well_known_resolver import WellKnownResolver
2021-09-27T15:54:47.4287856Z   File "/home/runner/work/synapse/synapse/synapse/http/federation/well_known_resolver.py", line 32, in <module>
2021-09-27T15:54:47.4288946Z     from synapse.util.caches.ttlcache import TTLCache
2021-09-27T15:54:47.4290319Z   File "/home/runner/work/synapse/synapse/synapse/util/caches/ttlcache.py", line 163, in <module>
2021-09-27T15:54:47.4291158Z     class _CacheEntry(Generic[KT, VT]):
2021-09-27T15:54:47.4309820Z   File "/home/runner/work/synapse/synapse/.tox/py/lib/python3.6/site-packages/attr/_make.py", line 1559, in wrap
2021-09-27T15:54:47.4310783Z     return builder.build_class()
2021-09-27T15:54:47.4313286Z   File "/home/runner/work/synapse/synapse/.tox/py/lib/python3.6/site-packages/attr/_make.py", line 730, in build_class
2021-09-27T15:54:47.4314208Z     return self._create_slots_class()
2021-09-27T15:54:47.4315773Z   File "/home/runner/work/synapse/synapse/.tox/py/lib/python3.6/site-packages/attr/_make.py", line 847, in _create_slots_class
2021-09-27T15:54:47.4316784Z     cls = type(self._cls)(self._cls.__name__, self._cls.__bases__, cd)
2021-09-27T15:54:47.4317678Z   File "/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/typing.py", line 948, in __new__
2021-09-27T15:54:47.4318610Z     raise TypeError("Cannot inherit from plain Generic")
2021-09-27T15:54:47.4319430Z builtins.TypeError: Cannot inherit from plain Generic
2021-09-27T15:54:47.4319918Z 
2021-09-27T15:54:47.4320688Z tests.rest.client.test_third_party_rules
```
</details>


Looks like https://github.com/python-attrs/attrs/issues/313

**Comment by DMRobertson** [comment]
CI happy now. Any objections @erikjohnston and @squahtx ?

**Comment by squahtx** [comment]
> CI happy now. Any objections @erikjohnston and @squahtx ?

No objections here

**Review by erikjohnston — APPROVED** [review]
